File: app/src/main/java/app/amphora/gamesession/input/GraphemeCounter.java
package app.amphora.gamesession.input;

import java.text.BreakIterator;
import java.util.Locale;

/** Counts user-visible character boundaries for IME deletion forwarding. */
public final class GraphemeCounter {
  private GraphemeCounter() {}

  public static int count(CharSequence text) {
    if (text == null || text.length() == 0) return 0;
    BreakIterator iterator = BreakIterator.getCharacterInstance(Locale.ROOT);
    iterator.setText(text.toString());
    int count = 0;
    for (int boundary = iterator.first();
        boundary != BreakIterator.DONE;
        boundary = iterator.next()) {
      if (boundary > 0) count++;
    }
    return count;
  }
}

File: app/src/main/kotlin/app/amphora/gamesession/input/GraphemeCounter.kt
package app.amphora.gamesession.input

import java.text.BreakIterator
import java.util.Locale

/** Counts user-visible character boundaries for IME deletion forwarding. */
object GraphemeCounter {
    @JvmStatic
    fun count(text: CharSequence?): Int {
        if (text.isNullOrEmpty()) return 0

        val iterator = BreakIterator.getCharacterInstance(Locale.ROOT)
        iterator.setText(text.toString())
        var count = 0
        var boundary = iterator.first()
        while (boundary != BreakIterator.DONE) {
            if (boundary > 0) count++
            boundary = iterator.next()
        }
        return count
    }
}
